### PR TITLE
[Eager] [Bitwise-Equivalence] Implement an INNER_TREE sum reduction (#182986)

### PR DESCRIPTION
**Pull request by njriasan**
Summary:

Adds the triton-style inner tree reduction kernel for sum in the caffe2/ATen CUDA backend, including partitioned launch variants, tests, and the output/input offset handling from D104676317 so TensorIterator output strides and strided outer input rows are preserved.

Test Plan:
arc lint -a fbcode/caffe2/aten/src/ATen/native/cuda/ReduceSumProdKernel.cu xplat/caffe2/aten/src/ATen/native/cuda/ReduceSumProdKernel.cu fbcode/caffe2/test/test_reductions.py xplat/caffe2/test/test_reductions.py

buck2 test fbcode//caffe2/test:torch_cuda_cuda -- TestInnerTreeSum

Differential Revision: D100024904

### CODE REVIEW

File: aten/src/ATen/native/cuda/ReduceSumProdKernel.cu [hunk 1]
```diff
@@ -7,12 +7,630 @@
...
+  const out_t* row_partials = partials + row_idx * num_batches;
+  out_t sum = row_partials[0];
+  for (int b = 1; b < num_batches; b++) {
+    sum = sum + row_partials[b];
+  }
+  output[row_idx] = sum;
+}
+
+static bool use_inner_tree_reduction() {
+  static const bool disabled = (std::getenv("PYTORCH_SUM_LEGACY") != nullptr);
```
Comment (njriasan) [inline aten/src/ATen/native/cuda/ReduceSumProdKernel.cu]: This will default to False after CI runs.

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Review by jansel — CHANGES_REQUESTED** [review]
fix test failures

**Comment by jansel** [comment]
@claude explain those CI failures, are they related?

**Comment by njriasan** [comment]
CI is passing now. I'll regenerate the performance numbers and update the config to disable the eager change by default.

**Comment by jansel** [comment]
@claude review these changes

**Comment by njriasan** [comment]
Performance Data:
H100 Data: https://gist.github.com/njriasan/df3025633042056be2d9ebb14817d0bc 
GB200 Data: https://gist.github.com/njriasan/f3217e39ddbc1ca7a94d7b764e7eee3e 
H100 LLM Analysis: https://gist.github.com/njriasan/75d8033900975be3eb40b32637b60599 
GB200 LLM Analysis: https://gist.github.com/njriasan/f3beb77118930b79ee62358b8649d698

**Comment by njriasan** [comment]
@claude review these changes

**Comment by njriasan** [comment]
> The "bitwise" test never exercises reduction ordering

This is a false positive. The reduction is done in fp32 so they actually should all have the same values. The upcast to f32 + same reduction ordering means its the same result (desired).

> try_inner_tree_reduction is still enabled by default

Waiting for a green CI + resolved issues before showing this.


I'll fix the others.

**Comment by njriasan** [comment]
@claude review these changes

**Comment by njriasan** [comment]
@claude review these changes

**Comment by njriasan** [comment]
These all look like reasonable leftovers. @jansel @eellison let me know if any of these seem high priority (aside from disabling the inner tree by default once CI passes and you both approve).

**Review by jansel — DISMISSED** [review]

File: aten/src/ATen/native/cuda/PersistentSoftmax.cuh [hunk 2]
```diff
@@ -142,7 +132,7 @@ __global__ void softmax_warp_forward(output_t *dst, const input_t *src, int batc
             }
         }
     }
-    warp_reduce<acc_t, WARP_BATCH, WARP_SIZE, Max>(max_value);
+    at::native::warp_reduce<acc_t, WARP_BATCH, WARP_SIZE, WarpReduceDirection::DESCENDING>(max_value, MaxOp{});
```
Comment (eellison) [inline aten/src/ATen/native/cuda/PersistentSoftmax.cuh]: max is invariant to reduction order so you dont necessarily need to update this...

**Review by eellison — COMMENTED** [review]
cc @slayton58 would you mind reviewing this ? it would be great to get eyes on this from someone with more cuda experience

File: aten/src/ATen/native/cuda/PersistentSoftmax.cuh (hunk 2 above)
Comment (njriasan) [reply aten/src/ATen/native/cuda/PersistentSoftmax.cuh]: Agreed. I believe this was to get a consistent single definition.

**Comment by njriasan** [comment]
The CI failures are hash inconsistencies related to the newly added tests. I will fix these ASAP.

File: aten/src/ATen/native/cuda/ReduceSumProdKernel.cu [hunk 3]
```diff
@@ -7,12 +7,618 @@
...
+    typename out_t,
+    int vec_size,
+    typename OutputCalculator>
+bool launch_multirow(
+    const scalar_t* input_ptr, out_t* output_ptr,
+    int64_t inputs_per_output, int64_t num_outputs,
+    OutputCalculator output_calc,
+    cudaStream_t stream) {
+  constexpr int kThreadsPerBlock = 128;
+  const int grid_size = static_cast<int>(
```
Comment (slayton58) [inline aten/src/ATen/native/cuda/ReduceSumProdKernel.cu]: Should check grid for very large `num_output: int64_t` here.

File: aten/src/ATen/native/cuda/ReduceSumProdKernel.cu [hunk 4]
```diff
@@ -7,12 +7,618 @@
...
+    const scalar_t* input_ptr, out_t* output_ptr,
+    int64_t inputs_per_output, int64_t num_outputs,
+    const InnerTreeLaunchParams& p,
+    OutputCalculator output_calc,
+    cudaStream_t stream) {
+  const int total_warps = p.num_warps * p.rows_per_block;
+  const dim3 block(C10_WARP_SIZE, total_warps);
+  const int shared_mem = (p.num_warps > 1)
+      ? static_cast<int>(sizeof(acc_t) * total_warps) : 0;
+  const dim3 grid((num_outputs + p.rows_per_block - 1) / p.rows_per_block);
```
Comment (slayton58) [inline aten/src/ATen/native/cuda/ReduceSumProdKernel.cu]: Same issue here

File: aten/src/ATen/native/cuda/ReduceSumProdKernel.cu [hunk 5]
```diff
@@ -7,12 +7,618 @@
...
+    cudaStream_t stream) {
+  const dim3 block(C10_WARP_SIZE, p.num_warps);
+  const int shared_mem = (p.num_warps > 1)
+      ? static_cast<int>(sizeof(acc_t) * p.num_warps) : 0;
+
+  auto& allocator = *c10::cuda::CUDACachingAllocator::get();
+  auto partials_storage = allocator.allocate(num_outputs * p.num_batches * sizeof(out_t));
+  auto* partials_ptr = static_cast<out_t*>(partials_storage.get());
+
+  const dim3 grid(num_outputs * p.num_batches);
```
Comment (slayton58) [inline aten/src/ATen/native/cuda/ReduceSumProdKernel.cu]: Same issue here

**Comment by slayton58** [comment]
The kernels themselves look ok to me - I can't guarantee no issues, but I don't see any, nor has claude flagged up anything concerning.

There is however missing size magnitude guard that should be added - repro & explanation per Claude:

> Missing magnitude guard: int truncation of the reduced length silently returns wrong results. `try_inner_tree_reduction` gates only on layout, never on size. `inputs_per_output` is int64_t but is truncated to int in `compute_inner_tree_params` (`const int n = static_cast<int>(inputs_per_output)`), in the device kernels' remaining/num_batches casts, and in `launch_multirow`'s int inputs_per_output. For a `reduced dim > INT_MAX` the cast goes negative, and the per-lane bound check `base + i < inputs_per_output` compares a corrupted int base against the true int64, so it doesn't catch it. The path is on by default, so any sum over a 2B+ contiguous dim is affected.
```
  import torch
  N = 2**31 + 1024                     # reduced length > INT_MAX
  x = torch.zeros(1, N, device="cuda", dtype=torch.float32)
  x[0, :500] = 1.0
  x[0, N-500:] = 1.0                   # true sum = 1000.0 (exact in fp32)
  print(x.sum(dim=1).item())           # -> 0.0  (PYTORCH_SUM_LEGACY=1 -> 1000.0)

#  inner-tree (default):    0.0      WRONG
#  PYTORCH_SUM_LEGACY=1:    1000.0    correct
#  inner-tree, N=1e5:       1000.0    correct (bug is magnitude-gated)
```

There also looks to be some missing checks on CUDA launch dimensions being too large - I've added those inline.

File: aten/src/ATen/native/cuda/ReduceSumProdKernel.cu [hunk 6]
```diff
@@ -7,12 +7,618 @@
...
+    inner_tree_reduction_looped_kernel<scalar_t, acc_t, out_t, vec_size, max_depth, wpr>
+        <<<grid, block, shared_mem, stream>>>(
+            input_ptr, output_ptr, inputs_per_output, num_outputs,
+            p.batch_total_elements, output_calc);
+  };
+
+  auto dispatch_depth = [&]<int wpr>() {
+    #define LAUNCH_CASE(D) case D: do_launch.template operator()<D, wpr>(); break
+    switch (p.depth) {
+      LAUNCH_CASE(1);  LAUNCH_CASE(2);  LAUNCH_CASE(3);  LAUNCH_CASE(4);
```
Comment (ngimel) [inline aten/src/ATen/native/cuda/ReduceSumProdKernel.cu]: this would drastically increase the number of templates we instantitate

**Comment by ngimel** [comment]
This is a sweeping change to one of the most commonly used pytorch operators. TBH, I don't think we should be doing this unless we have very good reasons.

**Comment by njriasan** [comment]
@claude review these changes. Please note its required to move this logic to CuteDSL.

**Comment by njriasan** [comment]
I will apply this feedback. However the `Compile-cache pressure` change will require me to evaluate with benchmarking as the immediate next step first.

**Comment by njriasan** [comment]
@claude review these changes. Note I have deferred the `Compile-cache pressure` until after I have recaptured performance measurements.

**Comment by njriasan** [comment]
@claude Review this code and verify the CI failures are unrelated

**Comment by njriasan** [comment]
@eellison @jansel @ngimel This is moved to CuteDSL and working now. The only remaining feedback to apply should be the decision to reduce "Compile-cache pressure". Other than the startup cost of hitting a new path with an initial compilation, can this have side-effects on other kernels (e.g. a max cache side) or will this be isolated behind the flag.

I will turn the flag off by default now that I have confirmed CI is working.

**Comment by slayton58** [comment]
@claude inspect the cuteDSL kernel code in this PR and suggest areas where existing cuteDSL abstractions and techniques could be used - pay special attention to vectorized loads / stores

**Comment by slayton58** [comment]
I would pay special attention to the host-side overhead of this override, especially with the 2-kernel versions of the code - it can easily make DSL code noticeably slower end-to-end even if the underlying kernel is faster.

I do also have a prototype implementation of all reduction operators in aten, with work in progress on pointwise -- it might be worth comparing notes to see if there's some common ground / kernels we can use.

**Comment by slayton58** [comment]
claude suggestions look reasonable to me - I don't remember seeing benchmarks, but aten's reductions are pretty well optimized and you'll need heavy vectorization to be competitive

**Comment by njriasan** [comment]
Here is the associated data from Karthick's benchmark on a GB200. I constructed two benchmarks. One where I just ran the kernel as is, shown below:

strict_numerics_row_reduction_gbps_vs_n

And one where I took the same kernels and implemented Cudagraphs supported (with a circular buffer on repeated measurements to avoid excessive cache hits).

strict_numerics_row_reduction_cudagraph_gbps_vs_n


Based on this result I think there are 3 real issues:
- Kernel launch overhead appears to be much greater with the CuteDSL version.
- Large M appears to be a regression at all N.
- Large N appears to be a regression for all M.

I believe at least the last 2 are possible to resolve based on prior experiments/optimization stacks, but I think they should be a followup since this will not be enabled by default.

**Review by karthickai — APPROVED** [review]
